gcs/client.go

```diff
@@ -0,0 +1,61 @@
+package gcs
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	gcStorage "cloud.google.com/go/storage"
+)
+
+type Client interface {
+	Read(ctx context.Context) ([]byte, error)
+	Write(ctx context.Context, p []byte) error
+}
+
+type Adapter struct {
+	config Config
+	client *gcStorage.Client
+}
+
+func (a Adapter) Read(ctx context.Context) ([]byte, error) {
+	r, err := a.client.Bucket(a.config.Bucket).Object(a.config.Name).NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	body, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading from gcs://%s/%s: %w", a.config.Bucket, a.config.Name, err)
+	}
+	return body, nil
+}
+
+func (a Adapter) Write(ctx context.Context, p []byte) error {
+	w := a.client.Bucket(a.config.Bucket).Object(a.config.Name).NewWriter(ctx)
+	_, err := w.Write(p)
+
+	if err != nil {
+		return fmt.Errorf("failed writing to gcs://%s/%s: %w", a.config.Bucket, a.config.Name, err)
+	}
+	return w.Close()
```

Comment:
I think we need to use `defer w.Close()` to close the Writer on error. Is there any reason to implement it differently than Read?

Reply:
It is because: 

1. We do not care the error returned by `r.Close()` because `io.readAll(r)` has already handled its responsibility. In other words, after `io.readAll(r)` worked properlly, we do not care that `r.Close()` returns error or not.
2. We do care the error returned by `w.Close()`, because it may mean that we failed to write/flush data to storage (GCS in this case). Even when `w.Write(...)` completed successfully, it may buffer the written data internally so we cannot say it is successfully saved onto the storage.

For reference, [a sample code in the official document](https://pkg.go.dev/cloud.google.com/go/storage@v1.25.0#hdr-Objects) handles error in the same manner:

```go
// Write some text to obj. This will either create the object or overwrite whatever is there already.
if _, err := fmt.Fprintf(w, "This object contains text.\n"); err != nil {
    // TODO: Handle error.
}
// Close, just like writing a file.
if err := w.Close(); err != nil {
    // TODO: Handle error.
}

// Read it back.
r, err := obj.NewReader(ctx)
if err != nil {
    // TODO: Handle error.
}
defer r.Close()
```

Reply:
Got it.

Though it’s technically possible to return an error from a defer function by using a named return value in general, I don’t have much knowledge about the GCS library implementation. It’s better to follow an official example unless we hit an actual issue.